redis/lookup_test.go

```diff
@@ -0,0 +1,146 @@
+package redis
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/letsencrypt/boulder/cmd"
+	blog "github.com/letsencrypt/boulder/log"
+	"github.com/letsencrypt/boulder/metrics"
+	"github.com/letsencrypt/boulder/test"
+
+	"github.com/redis/go-redis/v9"
+)
+
+func newTestRedisRing() *redis.Ring {
+	CACertFile := "../test/redis-tls/minica.pem"
+	CertFile := "../test/redis-tls/boulder/cert.pem"
+	KeyFile := "../test/redis-tls/boulder/key.pem"
+	tlsConfig := cmd.TLSConfig{
+		CACertFile: CACertFile,
+		CertFile:   CertFile,
+		KeyFile:    KeyFile,
+	}
+	tlsConfig2, err := tlsConfig.Load(metrics.NoopRegisterer)
+	if err != nil {
+		panic(err)
+	}
+
+	client := redis.NewRing(&redis.RingOptions{
+		Username:  "unittest-rw",
+		Password:  "824968fa490f4ecec1e52d5e34916bdb60d45f8d",
+		TLSConfig: tlsConfig2,
+	})
+	return client
+}
+
+func Test_Lookup(t *testing.T) {
+	t.Parallel()
+
+	logger := blog.NewMock()
+	ring := newTestRedisRing()
+
+	lookup := NewLookup([]cmd.ServiceDomain{
+		{
+			Service: "redisratelimits",
+			Domain:  "service.consul",
+		},
+	},
+		"consul.service.consul",
+		250*time.Millisecond,
+		ring,
+		logger,
+	)
+
+	testCtx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	lookup.Start(testCtx)
+
+	// The Consul service entry for 'redisratelimits' is configured to return
+	// two SRV targets. We should only have two shards in the ring.
+	test.Assert(t, ring.Len() == 2, "Expected 2 shards in the ring")
+
+	// Ensure we can reach both shards using the PING command.
+	err := ring.ForEachShard(testCtx, func(ctx context.Context, shard *redis.Client) error {
+		return shard.Ping(ctx).Err()
+	})
+	test.AssertNotError(t, err, "Expected PING to succeed for both shards")
+
+	// Drop both Shards from the ring.
+	ring.SetAddrs(map[string]string{})
+	test.Assert(t, ring.Len() == 0, "Expected 0 shards in the ring")
+
+	// Sleep 300ms to allow the periodic lookup to run.
+	time.Sleep(300 * time.Millisecond)
```

Comment:
Tests should avoid sleeping whenever possible.

The tests in this file are all "TestLookup", but Lookup is the struct. It makes these feel almost like integration tests. Instead, I think most of these tests would be restricted to a single call to lookup.Shards, and then either making assertions about its return value, or making assertions about the state of the ring (if you go with my comment above suggesting integrating the `ring.SetAddrs()` call into `Shards()`). None of those tests should have any need to sleep.

Then there can be just one additional test over `Start()`, which sets the update frequency to something absurd like 1ms, sleeps for just 5ms, and asserts that it has run at least twice (perhaps by inserting a mock redis Ring which records how many times SetAddrs is called).

Reply:
Should the test advance a fake clock instead of sleeping?

Reply:
Unfortunately that doesn't work for tickers: the clock package that we use (and the fake clock that it provides for testing) doesn't have support for triggering tickers when the clock advances the right amount. There are other clock packages that do have such support, but it's a bit flaky and they're missing other features and we'd have to do a big migration to use them.